File: code/0008. String to Integer (atoi)/solution1.cpp
class Solution {
public:
    int myAtoi(string s) {
        int i = 0, len = s.length();
        int sign = 1;
        long result = 0;

        // 1. Read in and ignore any leading whitespace.
        while (i < len && s[i] == ' ') {
            i++;
        }

        // 2. Check if the next character (if not already
        // at the end of the string) is '-' or '+'
        if (i < len && (s[i] == '-' || s[i] == '+')) {
            sign = s[i] == '-' ? -1 : 1;
            i++;
        }

        // 3. Read in next the characters until the next
        // non-digit character or the end of the input is reached.
        while (i < len && isdigit(s[i])) {

            // 4. Convert these digits into an integer
            int digit = s[i] - '0';
            result = result * 10 + digit;

            // 5. If the integer is out of the 32-bit signed
            // integer range [-2^31, 2^31 - 1], then clamp
            // the integer so that it remains in the range.
            if (result * sign > INT_MAX) {
                return INT_MAX;
            }
            if (result * sign < INT_MIN) {
                return INT_MIN;
            }
            i++;
        }

        // 6. Return the integer as the final result.
        return result * sign;
    }
};

File: code/1100. Find K-Length Substrings With No Repeated Characters/solution1.java
class Solution {
    public int numKLenSubstrNoRepeats(String s, int k) {
        int n = s.length(), result = 0;
        if (n < k) {
            return 0;
        }
        
        Map<Character, Integer> hashmap = new HashMap<>();
        int len = 0, start = 0;
        for (int i = 0; i < n; i++) {
            char c = s.charAt(i);
            if (!hashmap.containsKey(c) || hashmap.get(c) < start) {
                len++;
            }
            else {
                len = i - hashmap.get(c);
                start = hashmap.get(c) + 1;
            }
            hashmap.put(c, i);
            
            if (len >= k) {
                result++;
            }
        }
        
        return result;
    }
}